### PR TITLE
Replace Vec<u8> -> &[u8] - Closes #36

### LINKED ISSUE #36: Replace Vec<u8> with slices &[u8] as function parameters
### Description

- if a function has slice as parameter, it could accept both slices or Vectors as its argument
- Check if any function could accept a fixed size array as a parameter instead of a slice (eg. `&[u8; 32]` instead of `&[u8]`). In this case a compiler could verify if a function call has a valid value as argument.

### Acceptance Criteria

- functions have slices `&[u8]` as parameters instead of vectors `Vec<u8>`

### Additional Information

- only perform this two actions on a places in the code where and if it makes sense

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #36

### How was it solved?

Convert `Vec<u8>` -> `&[u8]` as arguments, if it is better option.

### How was it tested?

All tests passed.

### CODE REVIEW

**Review by shuse2 — COMMENTED** [review]
Generally, if we take as slice and immediately convert back to vector, does it make sense to change it? it seems like unnecessary additional computation

**Comment by matjazv** [comment]
> Generally, if we take as slice and immediately convert back to vector, does it make sense to change it? it seems like unnecessary additional computation

That's true.

But just as an example, if in `codec.rs` file, `struct Reader` is changed to:

> pub struct Reader<'a> {
>     index: usize,
>     end: usize,
>     data: &'a [u8],
> }

than function Reader::new can also be change to:

> pub fn new(data: &'a [u8]) -> Self {
>         let length = data.len();
>         Self {
>             data,
>             index: 0,
>             end: length,
>         }
>     }

and no additional `data.to_vec()` is needed.

@hrmhatef Try to look for the same opportunities also in other places in the code.

**Comment by hrmhatef** [comment]
> Generally, if we take as slice and immediately convert back to vector, does it make sense to change it? it seems like unnecessary additional computation



> > Generally, if we take as slice and immediately convert back to vector, does it make sense to change it? it seems like unnecessary additional computation
> 
> That's true.
> 
> But just as an example, if in `codec.rs` file, `struct Reader` is changed to:
> 
> ```
> pub struct Reader<'a> {
>     index: usize,
>     end: usize,
>     data: &'a [u8],
> }
> ```
> 
> than function Reader::new can also be change to:
> 
> ```
> pub fn new(data: &'a [u8]) -> Self {
>     let length = data.len();
>     Self {
>         data,
>         index: 0,
>         end: length,
>     }
> }
> ```
> 
> and no additional `data.to_vec()` is needed.
> 
> @hrmhatef Try to look for the same opportunities also in other places in the code.


If you are agree with all changes of this PR please approve it.
I am working on structs in my current task #40 so I will change structs like Matjaz mentioned.

File: src/state_writer.rs [hunk 1]
```diff
@@ -41,20 +41,19 @@ pub struct StateWriter {
...
     }
 
-    fn new_existing(val: Vec<u8>) -> Self {
-        let init = val.clone();
+    fn new_existing(val: &[u8]) -> Self {
         Self {
-            init: Some(init),
-            value: val,
+            init: Some(val.to_vec()),
+            value: val.to_vec(),
```
Comment (matjazv) [inline src/state_writer.rs:56]: Better to define `let val = val.to_vec()` outside of struct to not need to call `to_vec()` two times.

**Review by matjazv — CHANGES_REQUESTED** [review]

File: src/state_writer.rs (hunk 1 above)
Comment (hrmhatef) [reply src/state_writer.rs:56]: I think, it is not possible without extra `to_vec` or `clone`, because both `init` and `value` need to move `val`.

**Review by matjazv — APPROVED** [review]

**Review by shuse2 — APPROVED** [review]

**Pull request merged**